spec/lib/two_factor_authentication/models/two_factor_authenticatable_spec.rb

```diff
@@ -44,33 +44,76 @@
       end
     end
 
-    it_behaves_like 'otp_code', GuestUser.new
-    it_behaves_like 'otp_code', EncryptedUser.new
+    it_behaves_like 'generate_totp_code', GuestUser.new
+    it_behaves_like 'generate_totp_code', EncryptedUser.new
   end
 
-  describe '#authenticate_otp' do
-    shared_examples 'authenticate_otp' do |instance|
+  describe '#create_direct_otp' do
+    let(:instance) { build_guest_user }
+
+    it 'creates an otp_code' do
+      expect(instance.otp_code).to be_nil
+      instance.create_direct_otp
+      expect(instance.otp_code).not_to be_nil
+    end
+
+    it 'honors .direct_otp_length' do
+      expect(instance.class).to receive(:direct_otp_length).and_return(10)
+      instance.create_direct_otp
+      expect(instance.otp_code.length).to equal(10)
+
+      expect(instance.class).to receive(:direct_otp_length).and_return(6)
+      instance.create_direct_otp
+      expect(instance.otp_code.length).to equal(6)
+    end
+  end
+
+  describe '#authenticate_direct_otp' do
+    let(:instance) { build_guest_user }
+    it 'fails if no direct_otp has been set' do
+      expect(instance.authenticate_direct_otp('12345')).to eq(false)
+    end
+
+    context 'after generating an OTP' do
+      before :each do
+        instance.create_direct_otp
+      end
+
+      it 'authenticates with correct OTP' do
+        Timecop.freeze(Time.now + instance.class.direct_otp_valid_for - 1.second)
+        expect(instance.authenticate_direct_otp(instance.direct_otp)).to eq(true)
+      end
+
+      it 'rejects expired OTP' do
+        Timecop.freeze(Time.now + instance.class.direct_otp_valid_for + 1.second)
+        expect(instance.authenticate_direct_otp(instance.direct_otp)).to eq(false)
+      end
+    end
+  end
+
```

Comment:
I would also add a test for the scenario where a direct OTP exists, but the one passed in to the method does not match.